### ISSUE TITLE
Is there a way to know that the request is read-only, before calling `.end()`.

### ISSUE THREAD

**Issue by etnbrd** [sindresorhus/got#1927]
#### What would you like to discuss?

When reading the doc, I see that when using [streams](https://www.npmjs.com/package/got#streams), it is required to call the `.end()` method to send the request.
Moreover, if `options` contain a [`body`](https://www.npmjs.com/package/got#body) (or `form`, or `json`), then the response stream will be read-only.
> Note #2: If you provide this option, got.stream() will be read-only.

And if the `.end()` method is called while the stream is read-only, [it throws](https://github.com/sindresorhus/got/blob/49c605ca85fa35b7c1b6819219cb7bd5209b4085/source/core/index.ts#L556-L558).

Is there a way to know that the returned stream is read-only?
(Other than catching this error, or inferring the state of the returned stream based on the passed options)

I would expect the following to work, but apparently `writable` is always true.
```typescript
if (requestStream instanceof stream.Writable && requestStream.writable) {
  requestStream.end()
}
```

I tried to look up in the documentation and existing issues, but couldn't find anything.
Sorry if I missed it.

Also, I'd happy to try to implement something to make the `requestStream.writable` `false` when it's read-only.

#### Checklist

- [x] I have read the documentation.

**Comment by szmarczak** [comment]
I think you should be able to compare the `.end` function like this: `requestStream.end === stream.Writable.prototype.end`.

The behavior of automatically ending the stream has been here before I joined Got (v8.0.0):
https://github.com/sindresorhus/got/blob/570ff68c4853beda669ed597cd97f5acc24b9f60/index.js#L412

I agree that this solution is not optimal, it has led to this:
https://github.com/sindresorhus/got/blob/0f9f2b83b77710f2dc08c2a6bce1c78ba8d46760/source/create.ts#L59

We can enforce `.end()` to always be called, but that is a breaking change.

**Comment by sindresorhus** [comment]
Yeah, I agree, we should improve this. We could possible add an option in the meantime as v13 will probably not happen for a long time. The benefit of an option is that we get to see whether the change has any downsides before we make it the default.

**Comment by etnbrd** [comment]
Indeed, the following workaround works, thanks @szmarczak.
```typescript
if (requestStream.end === stream.Writable.prototype.end) {
  requestStream.end()
}
```

I'll close this issue. Feel free to re-open it if you want to use it to track the suggested improvement.
